src/platform_storage_api/admission_controller/api.py

```diff
@@ -0,0 +1,219 @@
+
+LABEL_APOLO_ORG_NAME = "platform.apolo.us/org"
+LABEL_APOLO_PROJECT_NAME = "platform.apolo.us/project"
+LABEL_PLATFORM_STORAGE_POD = "platform-storage"
+
+INJECTED_VOLUME_NAME_PREFIX = "storage-auto-injected-volume"
+
+
+def create_injection_volume_name() -> str:
+    """Creates a random volume name"""
+    return f"{INJECTED_VOLUME_NAME_PREFIX}-{uuid4().hex}"
+
+
+class AdmissionReviewPatchType(str, Enum):
+    JSON = "JSONPatch"
+
+
+@dataclasses.dataclass
+class AdmissionReviewResponse:
+    uid: str
+    allowed: bool = True
+    patch: Optional[list[dict[str, Any]]] = None
+    patch_type: AdmissionReviewPatchType = AdmissionReviewPatchType.JSON
+
+    def add_patch(self, path: str, value: Any) -> None:
+        if self.patch is None:
+            self.patch = []
+
+        self.patch.append({
+            "op": "add",
+            "path": path,
+            "value": value,
+        })
+
+    def to_dict(self) -> dict[str, Any]:
+        patch: Optional[str] = None
+
+        if self.patch is not None:
+            # convert patch changes to a b64
+            dumped = json.dumps(self.patch).encode()
+            patch = base64.b64encode(dumped).decode()
+
+        return {
+            "apiVersion": "admission.k8s.io/v1",
+            "kind": "AdmissionReview",
+            "response": {
+                "uid": self.uid,
+                "allowed": self.allowed,
+                "patch": patch,
+                "patchType": AdmissionReviewPatchType.JSON.value
+            }
+        }
+
+    def to_response(self) -> web.Response:
+        return web.json_response(self.to_dict())
+
+
+class AdmissionControllerApi:
+
+    def __init__(
+        self,
+        app: web.Application,
+        config: Config
+    ) -> None:
+
+        self._app = app
+        self._config = config
+
+    @property
+    def _volume_resolver(self) -> KubeVolumeResolver:
+        return self._app[VOLUME_RESOLVER_KEY]
+
+    def register(self, app: web.Application) -> None:
+        app.add_routes([
+            web.post("/mutate", self.handle_post_mutate)
+        ])
+
+    async def handle_post_mutate(self, request: web.Request) -> Any:
+        logger.info("mutate call")
+        payload: dict[str, Any] = await request.json()
+
+        uid = payload["request"]["uid"]
+        admission_review = AdmissionReviewResponse(uid=uid)
+
+        pod = payload["request"]["object"]
+        kind = pod.get("kind")
+
+        if kind != "Pod":
+            # not a pod creation request. early-exit
+            return admission_review.to_response()
+
+        metadata = pod.get("metadata", {})
+        labels = metadata.get("labels")
+
+        # let's check if this is a new platform storage POD
+        if (
+            labels.get("app") == LABEL_PLATFORM_STORAGE_POD and
+            labels.get("service") == LABEL_PLATFORM_STORAGE_POD
+        ):
+            return await self._handle_new_platform_storage_pod(
```

Comment:
I was thinking that we could mount all volumes to admission controller. In this case we won't need to listen to changes in platform storage pods, it will have the latest available volumes. We will only need to load volumes of the pod hosting this service during startup. wdyt?

Reply:
changed that